src/appleseed/foundation/meta/tests/test_imagefilewriter.cpp

```diff
@@ -0,0 +1,201 @@
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2019-2020 Chin-Ying Li, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.foundation headers.
+#include "foundation/image/color.h"
+#include "foundation/image/colorspace.h"
+#include "foundation/image/genericimagefilereader.h"
+#include "foundation/image/genericimagefilewriter.h"
+#include "foundation/image/image.h"
+#include "foundation/image/imageattributes.h"
+#include "foundation/image/pixel.h"
+#include "foundation/math/scalar.h"
+#include "foundation/utility/iostreamop.h"
+#include "foundation/utility/test.h"
+
+// Standard headers.
+#include <cassert>
+#include <cmath>
+#include <cstddef>
+#include <cstring>
+#include <exception>
+#include <memory>
+
+using namespace foundation;
+
+TEST_SUITE(Foundation_Image_ImageFileWriter)
+{
+    template<typename Color>
+    void write_image(const Color fill_color, const PixelFormat pixel_format, const size_t s, const std::string image_file_path)
+    {
+        size_t channels = fill_color.Components;
+        
+        std::unique_ptr<Image> image(new Image(s, s, s, s, channels, pixel_format));
+        image->clear(fill_color);
+        
+        GenericImageFileWriter writer(image_file_path.c_str());
+        writer.append_image(image.get());
+        writer.write();
+    }
+    
+    TEST_CASE(WriteBMP_CorrectlyWritesImagePixels)
+    {
+        const std::string image_file_path = "unit tests/outputs/test_imagefilewriter.bmp";
+        const size_t s = 16;
```

Comment:
Use `std::size_t` in new code (fix in all new code introduced in this PR).

Rename `s` to something clearer, maybe `ImageSize`?